File: parameters/calibration/bayesian_calibration.py

import numpy as np
from sklearn.covariance import log_likelihood
import statsmodels.api as sm
import scipy

from calibration_helpers import *

class BayesianEstimator:

    def __init__(self, X, Y, Y_true):
        self.X = X
        self.Y = Y
        self.Y_true = Y_true

        self.priors = None
        self.f = None
        self.posterior = None

    def compute_coeff(self):
        
        # Fit likelihood
        self.compute_f()

        self.sample_posterior()


    def compute_f(self):

        # print(self.Y[0,0], self.Y[0,1], self.Y[0,2])
        
        self.f = sm.nonparametric.KDEMultivariateConditional(
                    endog=self.Y,
                    exog=self.X, 
                    dep_type='c' * self.Y.shape[1],
                    indep_type='c' * self.X.shape[1],
                    bw='normal_reference'
                )

    def compute_likelihood(self, x):
        # likelihood = 1
        # for y in self.Y_true:
        #     # print(x)
        #     # print(y)
        #     a = self.f.pdf(endog_predict=y, exog_predict=x)
        #     # print(a)
        #     likelihood *= a
        # return likelihood

        x = np.repeat([x], len(self.Y_true), axis=0)

        return np.product(self.f.pdf(endog_predict=self.Y_true, 
                                     exog_predict=x))

    def checkbounds(self, x_new, bounds):
        for i,x in enumerate(x_new):
            if x < bounds[i][0] or x > bounds[i][1]:
                return True
        return False

    def sample_posterior(self):

        x_curr = np.array([0.05, 0.2, 0.005, 0.8, 0.02, 0.8, 0.5, 0.0])
        bounds = [(0.05, 0.05),
                  (0.2, 0.2),
                  (0.001, 0.01),
                  (0., 1.),
                  (0., 0.05),
                  (0.6, 1.),
                  (0., 1.),
                  (-1., 1.)]
        
        weights = 10 * np.array([0, 0, 0.001, .5, 0.01, .5, .7, .7])
        curr_lh = self.compute_likelihood(x_curr)

        all_x = []

        for _ in range(100):

            selec = np.zeros(len(x_curr))
            selec_idx = np.random.randint(2, len(x_curr))
            selec[selec_idx] = 1.

            epsilon = np.random.normal(0, 0.1, 8)
            x_new = x_curr + epsilon * weights * selec

            while self.checkbounds(x_new, bounds):
                epsilon = np.random.normal(0, 0.1, 8)
                x_new = x_curr + epsilon * weights * selec
            
            new_lh = self.compute_likelihood(x_new)
            print(new_lh)

            alpha = min(1, new_lh / curr_lh)
            print(alpha)

            if np.random.rand() < alpha:
                x_curr = x_new
                curr_lh = new_lh
                all_x.append(x_curr)
                # print(x_curr)

        all_x = np.array(all_x)
        res = {'κ_upper': all_x.T[2], 'ω': all_x.T[3], 'ϵ': all_x.T[4], 
               'α_cp': all_x.T[5], 'p_f': all_x.T[6], 'prog': all_x.T[7]}
        pd.DataFrame(res).to_csv('postdists.csv', index=False)


if __name__ == "__main__":

    inputpath = '../sensitivity/sensitivity_runs/input_data/'
    outputpath = 'calibrationdata/'
    n_threads = 4

    # Y_true = {
    #     'GDP_1st': 0.482,
    #     'GDP_2nd': 0.47,
    #     'U_1st': 0.06097,
    #     'em2030': 112.72,
    #     'em2030': 121.82,
    #     'em2040': 127.27
    # }

    # dep_vars = Y_true.keys()

    Y_true = pd.read_csv(outputpath + 'truedata.csv')[["dGDP", "em"]].to_numpy()

    X, Y = load_data(inputpath, outputpath, n_threads)

    estimator = BayesianEstimator(X, Y, Y_true)
    estimator.compute_coeff()

File: parameters/calibration/bayesian_calibration.jl
using PyCall
using ArgParse
using DataFrames
using CSV
using Random
using Tables

include("../../model/main.jl")

pysm = pyimport("statsmodels.api")


function run_calibration(
    ytrue::Matrix{Float64},
    outputpath::String
    )

    labels = ["κ_upper", "ω", "ϵ", "α_cp", "p_f", "prog"]
    xbounds = [[0.001, 0.01], [0.0, 1.0], [0.05, 0.5], [0.65, 0.95], [0.1, 1.0], [-1.0, 1.0]]
    
    # Define initial x and weights used for transformations
    xcurr = Float64[0.005227455, 0.429497195, 0.085293161, 0.808659944, 0.340361401, 0.116886424]
    # xcurr = Float64[0.005, 0.5, 0.2, 0.85, 0.3, 0.0]
    xweights = Float64[0.001, 0.1, 0.001, 0.1, 0.01, 0.01]

    # Get y and likelihood from initial x
    y = run_model(xcurr, labels)
    lhcurr = compute_likelihood(y, ytrue)

    filepath = outputpath * "bayes_sampled_params_thread.csv"
    savedata(0, lhcurr, xcurr, labels, filepath; append=false)

    print(lhcurr)

    for i in 1:100

        # Draw new set of parameter values x
        xnew = gen_x(xcurr, xweights, xbounds)
        ynew = run_model(xnew, labels)
        lhnew = compute_likelihood(ynew, ytrue)

        # Compute acceptance chance
        α = min(1, lhnew / lhcurr)

        println(xnew)
        println(lhnew)
        println(α)

        if Random.rand() < α
            # Save parameter set
            xcurr .= xnew
            lhcurr = lhnew
            savedata(i, lhcurr, xcurr, labels, filepath)
            # res = vcat([i, lhcurr], xcurr)
            # CSV.write(filepath, Tables.table(res))
        end
    end
end

function savedata(
    i::Int64,
    lhcurr::Float64,
    xcurr::Vector{Float64},
    labels::Vector{String},
    filepath::String;
    append::Bool=true
    )

    # Initialize dataframe that holds parameter values, write to CSV
    df = DataFrame(vcat(["sim_nr", "lh"], labels) .=> vcat([i, lhcurr], xcurr))
    CSV.write(filepath, df, delim=';', append=append)
end


function compute_likelihood(
    y::Matrix{Float64},
    ytrue::Matrix{Float64}
    )

    f = pysm.nonparametric.KDEMultivariate(
        data=y,
        var_type="ccc",
        bw="normal_reference"
    )
    likelihood = prod(f.pdf(ytrue))
    return likelihood
end


function gen_x(xcurr, xweights, xbounds)

    # Draw which parameter will be changed
    choice = zeros(length(xcurr))
    choice[rand(1:length(xcurr))] = 1

    # Draw stochastic shocks
    ε = Random.randn(length(xcurr)) .* xweights .* choice
    xnew = xcurr .+ ε

    # Make sure parameter values stay within their bounds
    while !inbounds(xnew, xbounds)
        ε = Random.randn(length(xcurr)) .* xweights .* choice
        xnew = xcurr .+ ε
    end
    return xnew
end


function inbounds(
    xnew, 
    xbounds
    )

    for (x, bound) in zip(xnew, xbounds)
        if x < bound[1] || x > bound[2]
            return false
        end
    end
    return true
end


function run_model(
    xcurr::Vector{Float64}, 
    labels::Vector{String}
    )

    changedparams = Dict(labels .=> xcurr)

    runoutput = run_simulation(
        changed_params=changedparams,
        full_output=false;
        threadnr=Threads.threadid()
    )

    return hcat(
            runoutput.GDP_growth[100:end], 
            100 .* runoutput.U[100:end], 
            runoutput.emissions_index[100:end]
           )
end

function main()

    inputpath = "../sensitivity/sensitivity_runs/input_data/"
    outputpath = "calibrationdata/"
    n_threads = 16
    run_nr = 9

    ytrue = Matrix(DataFrame(CSV.File(outputpath * "truedata.csv")))

    run_calibration(ytrue, outputpath)

end

main()